File: practices/chapter11_maze/python/ch_11_maze.py
from pprint import pprint


# 0 為可以走的, 1 為牆壁, 2為已走過的路徑, 3為標示不可走的
small_maze = [
  [1, 1, 1, 1, 1, 1],
  [1, 0, 1, 1, 1, 1],
  [1, 0, 1, 0, 1, 1],
  [1, 0, 0, 0, 1, 1],
  [1, 0, 1, 0, 0, 1],
  [1, 1, 1, 1, 1, 1],
]

big_maze = [
  [1, 1, 1, 1, 1, 1, 1, 1, 1],
  [1, 0, 1, 1, 1, 1, 1, 1, 1],
  [1, 0, 1, 1, 1, 1, 1, 1, 1],
  [1, 0, 1, 1, 0, 0, 0, 0, 1],
  [1, 0, 1, 1, 0, 0, 0, 0, 1],
  [1, 0, 0, 0, 0, 0, 1, 0, 1],
  [1, 0, 1, 0, 1, 0, 1, 0, 1],
  [1, 1, 0, 0, 1, 0, 1, 0, 1],
  [1, 1, 1, 1, 1, 1, 1, 1, 1],
]

directions_fns = [
  lambda x, y: (x - 1, y), # left
  lambda x, y: (x + 1, y), # right
  lambda x, y: (x, y - 1), # down
  lambda x, y: (x, y + 1), # up
]

def solve_maze(start_pos=[], goal_pos=[], maze=[]):
  maze_res = maze
  maze_res[start_pos[0]][start_pos[1]] = 2
  maze_path_stack = []
  maze_path_stack.append(start_pos)

  while len(maze_path_stack) > 0:
    current_pos = maze_path_stack[-1] # last position
    if current_pos[0] == goal_pos[0] and current_pos[1] == goal_pos[1]:
      print('Maze cleared!')
      return maze_res
    for direction in directions_fns:
      next_pos = direction(current_pos[0], current_pos[1])
      if maze_res[next_pos[0]][next_pos[1]] == 0:
        maze_path_stack.append(next_pos)
        maze_res[next_pos[0]][next_pos[1]] = 2
        break
    else:
      maze_res[current_pos[0]][current_pos[1]] = 3
      maze_path_stack.pop()

  print('Path not found')
  return maze_res

def main():
  solved_maze = solve_maze([1, 1], [4, 4], small_maze)
  pprint(solved_maze)

def convert_input_to_arr(input_str='1,1'):
  split_arr = input_str.split(',')
  res = list(map(lambda str: int(str), split_arr))
  return res

def custom_start_maze():
  # start_pos_str = '1,1'
  start_pos_str = input('Input start(e.g. 1,1): ')
  # goal_pos_str = '7,7'
  goal_pos_str = input('Input goal(e.g. 7,7): ')

  start_pos = convert_input_to_arr(start_pos_str)
  goal_pos = convert_input_to_arr(goal_pos_str)
  print(start_pos, goal_pos)
  solved_maze = solve_maze(start_pos, goal_pos, big_maze)
  pprint(solved_maze)

if __name__ == "__main__":
    # main()
    custom_start_maze()

File: practices/chapter11_maze/js/ch_11_maze.js
const {
  rl
} = require('../../lib/cmdReadline')

// 0 為可以走的, 1 為牆壁, 2為已走過的路徑, 3為標示不可走的
const smallMaze = [
  [1, 1, 1, 1, 1, 1],
  [1, 0, 1, 1, 1, 1],
  [1, 0, 1, 0, 1, 1],
  [1, 0, 0, 0, 1, 1],
  [1, 0, 1, 0, 0, 1],
  [1, 1, 1, 1, 1, 1],
]

const bigMaze = [
  [1, 1, 1, 1, 1, 1, 1, 1, 1],
  [1, 0, 1, 1, 1, 1, 1, 1, 1],
  [1, 0, 1, 1, 1, 0, 1, 1, 1],
  [1, 0, 1, 1, 0, 0, 0, 0, 1],
  [1, 0, 1, 1, 1, 0, 1, 0, 1],
  [1, 0, 0, 0, 0, 0, 1, 0, 1],
  [1, 0, 1, 0, 1, 0, 1, 0, 0],
  [1, 1, 0, 0, 1, 0, 1, 0, 1],
  [1, 1, 1, 1, 1, 1, 1, 1, 1],
] // [0, 1], [8, 6]

const directionFns = [
  (x, y) => [x - 1, y],
  (x, y) => [x + 1, y],
  (x, y) => [x, y - 1],
  (x, y) => [x, y + 1],
]

function solveMaze(maze=[]) {
  return (startPos, goalPos) => {
    let solvedRes = 'no-result'
    let solvedMaze = [...maze]
    let passedPathStack = []
    
    solvedMaze[startPos[0]][startPos[1]] = 2
    passedPathStack.push(startPos)

    while(passedPathStack.length > 0) {
      const currentPos = [...passedPathStack].pop()
      console.log(currentPos)
      if(currentPos[0] === goalPos[0] && currentPos[1] === goalPos[1]) {
        return ({
          solvedRes: 'solved',
          solvedMaze,
          passedPathStack,
        })
      }

      else if (!(currentPos[0] === goalPos[0] && currentPos[1] === goalPos[1])) {
        for (const directionFn of directionFns) {
          const nextPos = directionFn(currentPos[0], currentPos[1])
          // console.log(nextPos)
          const nextMazeStatus = solvedMaze[nextPos[0]][nextPos[1]]
          if(nextMazeStatus === 0) {
            passedPathStack.push(nextPos)
            solvedMaze[nextPos[0]][nextPos[1]] = 2
            break
          }
        }
      } else {
        solvedMaze[currentPos[0]][currentPos[1]] = 3
        passedPathStack.pop()
      }
    }

    return ({
      solvedRes,
      solvedMaze,
      passedPathStack,
    })
  }
}

function convertInputToArr(input_str='1,1') {
  const split_arr = input_str
    .split(',')
    .map(s => Number(s))
  return split_arr
}

function main() {
  let start = [1, 1]
  let goal = [8, 6]

  // rl.question('Input start(e.g. 1,1): ', ans => {
  //   start = convertInputToArr(ans)
  //   rl.question('Input goal(e.g. 7,7): ', ans => {
  //     goal = convertInputToArr(ans)
  //   })
  // })
  
  const solvedMazeRes = solveMaze(bigMaze)(start, goal)
  console.log(solvedMazeRes)

  rl.close()
}

main()